<a href="https://colab.research.google.com/github/HoagondeAmedegnato/Banks-Historical-Stock-Price-Portfolio/blob/main/01_Assets_Allocation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**BANKS HISTORICAL STOCK PRICE**

Data Source: https://www.kaggle.com/tomasmantero/banks-historical-stock-price

***KPI: Help Investors Make Important Decisions And Predict New Trends***

author = Hoagonde Amedegnato

email = amereine69@gmail.com

# **PORTFOLIO ASSETS ALLOCATION AND STATISTICAL DATA ANALYSIS**

- UNDERSTAND THE CONCEPT OF ASSET ALLOCATIONS:
- Asset allocation is an investment strategy that is used to alocate client's assets based on their risk tolerance, target returns and investment time span.
- The conventional wisdom is to subtract client's age from 100 to calculate the % of portolio that should be allocated to stocks
- For Example: If you're 30 years old, you should have 70% allocated to stocks
- If you are 75 years old, you should invest 25% in stocks (old retired seniors generally recommend a more stable low risk portfolio).

In [1]:
# Mount the drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Import Libraries
import pandas as pd
import plotly.express as px
from copy import copy
from scipy import stats
import matplotlib.pyplot as plt
import numpy as np
import plotly.figure_factory as ff
import plotly.graph_objects as go

import warnings
warnings.filterwarnings('ignore')

# *Load and Read Data*

In [3]:
BAC = pd.read_csv('/content/drive/MyDrive/Banks Stock Price/BAC (2).csv')
BAC

Date       High        Low  ...      Close      Volume  Adj Close
0     2006-01-03  47.180000  46.150002  ...  47.080002  16296700.0  34.811729
1     2006-01-04  47.240002  46.450001  ...  46.580002  17757900.0  34.442013
2     2006-01-05  46.830002  46.320000  ...  46.639999  14970700.0  34.486385
3     2006-01-06  46.910000  46.349998  ...  46.570000  12599800.0  34.434616
4     2006-01-09  46.970001  46.360001  ...  46.599998  15619400.0  34.456806
...          ...        ...        ...  ...        ...         ...        ...
3744  2020-11-16  28.000000  27.180000  ...  27.580000  57396700.0  27.580000
3745  2020-11-17  27.600000  27.010000  ...  27.549999  46402300.0  27.549999
3746  2020-11-18  27.930000  26.959999  ...  26.980000  54416700.0  26.980000
3747  2020-11-19  27.010000  26.650000  ...  26.980000  46423000.0  26.980000
3748  2020-11-20  26.950001  26.629999  ...  26.809999  40513900.0  26.809999

[3749 rows x 7 columns]

In [4]:
C = pd.read_csv('/content/drive/MyDrive/Banks Stock Price/C.csv')
C

Date        High         Low  ...       Close      Volume   Adj Close
0     2006-01-03  493.799988  481.100006  ...  492.899994   1537600.0  379.511047
1     2006-01-04  491.000000  483.500000  ...  483.799988   1870900.0  372.504486
2     2006-01-05  487.799988  484.000000  ...  486.200012   1143100.0  374.352448
3     2006-01-06  489.000000  482.000000  ...  486.200012   1370200.0  374.352448
4     2006-01-09  487.399994  483.000000  ...  483.899994   1680700.0  372.581421
...          ...         ...         ...  ...         ...         ...         ...
3744  2020-11-16   51.020000   49.860001  ...   50.400002  26699400.0   50.400002
3745  2020-11-17   50.930000   49.480000  ...   50.730000  18970000.0   50.730000
3746  2020-11-18   52.650002   51.009998  ...   51.049999  25824000.0   51.049999
3747  2020-11-19   52.320000   50.950001  ...   52.139999  20262400.0   52.139999
3748  2020-11-20   52.110001   51.189999  ...   51.650002  15471100.0   51.650002

[3749 rows x 7 columns]

In [5]:
GS = pd.read_csv('/content/drive/MyDrive/Banks Stock Price/GS.csv')
GS

Date        High         Low  ...       Close     Volume   Adj Close
0     2006-01-03  129.440002  124.230003  ...  128.869995  6188700.0  106.263367
1     2006-01-04  128.910004  126.379997  ...  127.089996  4861600.0  104.795609
2     2006-01-05  127.320000  125.610001  ...  127.040001  3717400.0  104.754387
3     2006-01-06  129.250000  127.290001  ...  128.839996  4319600.0  106.238571
4     2006-01-09  130.619995  128.000000  ...  130.389999  4723500.0  107.516731
...          ...         ...         ...  ...         ...        ...         ...
3744  2020-11-16  224.600006  220.029999  ...  222.380005  3004700.0  222.380005
3745  2020-11-17  224.850006  218.330002  ...  224.649994  2722900.0  224.649994
3746  2020-11-18  227.720001  223.509995  ...  224.300003  2856200.0  224.300003
3747  2020-11-19  224.940002  222.110001  ...  224.570007  1505200.0  224.570007
3748  2020-11-20  224.809998  222.630005  ...  223.350006  1848100.0  223.350006

[3749 rows x 7 columns]

In [6]:
JPM = pd.read_csv('/content/drive/MyDrive/Banks Stock Price/JPM.csv')
JPM

Date        High         Low  ...       Close      Volume   Adj Close
0     2006-01-03   40.360001   39.299999  ...   40.189999  12838600.0   27.013376
1     2006-01-04   40.139999   39.419998  ...   39.619999  13491500.0   26.857462
2     2006-01-05   39.810001   39.500000  ...   39.740002   8109400.0   26.938803
3     2006-01-06   40.240002   39.549999  ...   40.020000   7966900.0   27.128614
4     2006-01-09   40.720001   39.880001  ...   40.669998  16575200.0   27.569229
...          ...         ...         ...  ...         ...         ...         ...
3744  2020-11-16  118.610001  115.730003  ...  117.300003  19858600.0  117.300003
3745  2020-11-17  116.440002  114.580002  ...  116.110001  14553200.0  116.110001
3746  2020-11-18  118.610001  115.199997  ...  115.250000  12571200.0  115.250000
3747  2020-11-19  115.820000  113.559998  ...  115.559998  10202500.0  115.559998
3748  2020-11-20  115.650002  114.089996  ...  114.570000  12718900.0  114.570000

[3749 rows x 7 columns]

In [7]:
MS = pd.read_csv('/content/drive/MyDrive/Banks Stock Price/MS.csv')
MS

Date       High        Low  ...      Close      Volume  Adj Close
0     2006-01-03  58.490002  56.740002  ...  58.310001   5377000.0  36.867344
1     2006-01-04  59.279999  58.349998  ...  58.349998   7977800.0  36.892612
2     2006-01-05  58.590000  58.020000  ...  58.509998   5778000.0  36.993786
3     2006-01-06  58.849998  58.049999  ...  58.570000   6889800.0  37.031731
4     2006-01-09  59.290001  58.619999  ...  59.189999   4144500.0  37.423725
...          ...        ...        ...  ...        ...         ...        ...
3744  2020-11-16  58.139999  57.029999  ...  58.110001  12201400.0  58.110001
3745  2020-11-17  58.889999  57.200001  ...  58.810001   9599100.0  58.810001
3746  2020-11-18  60.130001  58.770000  ...  58.810001  11536700.0  58.810001
3747  2020-11-19  58.689999  57.860001  ...  58.630001   7220200.0  58.630001
3748  2020-11-20  58.740002  57.509998  ...  57.810001   7739900.0  57.810001

[3749 rows x 7 columns]

In [8]:
WFC = pd.read_csv('/content/drive/MyDrive/Banks Stock Price/WFC (1).csv')
WFC

Date       High        Low  ...      Close      Volume  Adj Close
0     2006-01-03  31.975000  31.195000  ...  31.900000  11016400.0  20.723480
1     2006-01-04  31.820000  31.365000  ...  31.530001  10870000.0  20.483118
2     2006-01-05  31.555000  31.309999  ...  31.495001  10158000.0  20.460379
3     2006-01-06  31.775000  31.385000  ...  31.680000   8403800.0  20.580564
4     2006-01-09  31.825001  31.555000  ...  31.674999   5619600.0  20.577309
...          ...        ...        ...  ...        ...         ...        ...
3744  2020-11-16  25.030001  24.389999  ...  24.900000  36954800.0  24.900000
3745  2020-11-17  25.150000  24.520000  ...  25.040001  29490400.0  25.040001
3746  2020-11-18  25.950001  25.059999  ...  25.059999  44318300.0  25.059999
3747  2020-11-19  26.219999  24.969999  ...  26.160000  44560000.0  26.160000
3748  2020-11-20  26.230000  25.459999  ...  25.480000  38425200.0  25.480000

[3749 rows x 7 columns]

In [9]:
# Let's create a list for all banks stocks 
tickers = ['BAC', 'C', 'GS', 'JPM', 'MS', 'WFC']

In [10]:
# Let's use pd.concat to concatenate all these bank stocks together into a single dataframe
bank_stocks = pd.concat([BAC, C, GS, JPM, MS, WFC], axis = 1, keys = tickers)
bank_stocks

BAC                        ...        WFC                       
            Date       High        Low  ...      Close      Volume  Adj Close
0     2006-01-03  47.180000  46.150002  ...  31.900000  11016400.0  20.723480
1     2006-01-04  47.240002  46.450001  ...  31.530001  10870000.0  20.483118
2     2006-01-05  46.830002  46.320000  ...  31.495001  10158000.0  20.460379
3     2006-01-06  46.910000  46.349998  ...  31.680000   8403800.0  20.580564
4     2006-01-09  46.970001  46.360001  ...  31.674999   5619600.0  20.577309
...          ...        ...        ...  ...        ...         ...        ...
3744  2020-11-16  28.000000  27.180000  ...  24.900000  36954800.0  24.900000
3745  2020-11-17  27.600000  27.010000  ...  25.040001  29490400.0  25.040001
3746  2020-11-18  27.930000  26.959999  ...  25.059999  44318300.0  25.059999
3747  2020-11-19  27.010000  26.650000  ...  26.160000  44560000.0  26.160000
3748  2020-11-20  26.950001  26.629999  ...  25.480000  38425200.0  25.480000

[3749 rows x 42 columns]

# *Create Bank Stocks Price (Close) DataFrame*

In [11]:
BAC = bank_stocks['BAC', 'Close']
BAC

0       47.080002
1       46.580002
2       46.639999
3       46.570000
4       46.599998
          ...    
3744    27.580000
3745    27.549999
3746    26.980000
3747    26.980000
3748    26.809999
Name: (BAC, Close), Length: 3749, dtype: float64

In [12]:
C = bank_stocks['C', 'Close']
C

0       492.899994
1       483.799988
2       486.200012
3       486.200012
4       483.899994
           ...    
3744     50.400002
3745     50.730000
3746     51.049999
3747     52.139999
3748     51.650002
Name: (C, Close), Length: 3749, dtype: float64

In [13]:
GS = bank_stocks['GS', 'Close']
GS

0       128.869995
1       127.089996
2       127.040001
3       128.839996
4       130.389999
           ...    
3744    222.380005
3745    224.649994
3746    224.300003
3747    224.570007
3748    223.350006
Name: (GS, Close), Length: 3749, dtype: float64

In [14]:
JPM = bank_stocks['JPM', 'Close']
JPM

0        40.189999
1        39.619999
2        39.740002
3        40.020000
4        40.669998
           ...    
3744    117.300003
3745    116.110001
3746    115.250000
3747    115.559998
3748    114.570000
Name: (JPM, Close), Length: 3749, dtype: float64

In [15]:
MS = bank_stocks['MS', 'Close']
MS

0       58.310001
1       58.349998
2       58.509998
3       58.570000
4       59.189999
          ...    
3744    58.110001
3745    58.810001
3746    58.810001
3747    58.630001
3748    57.810001
Name: (MS, Close), Length: 3749, dtype: float64

In [16]:
WFC = bank_stocks['WFC', 'Close']
WFC

0       31.900000
1       31.530001
2       31.495001
3       31.680000
4       31.674999
          ...    
3744    24.900000
3745    25.040001
3746    25.059999
3747    26.160000
3748    25.480000
Name: (WFC, Close), Length: 3749, dtype: float64

In [17]:
# Since all banks have the same Date, let's select the date column to concatenate 
DATE = bank_stocks['C', 'Date']
DATE

0       2006-01-03
1       2006-01-04
2       2006-01-05
3       2006-01-06
4       2006-01-09
           ...    
3744    2020-11-16
3745    2020-11-17
3746    2020-11-18
3747    2020-11-19
3748    2020-11-20
Name: (C, Date), Length: 3749, dtype: object

In [18]:
# Let's create a list for all banks stocks Price using close column
tickers_close = ['DATE', 'BAC', 'C', 'GS', 'JPM', 'MS', 'WFC']

In [19]:
# Let's use pd.concat to concatenate all these bank stocks close column together into a single dataframe
bank_stocks_close = pd.concat([DATE, BAC, C, GS, JPM, MS, WFC], axis = 1, keys = tickers_close)
bank_stocks_close

DATE        BAC           C  ...         JPM         MS        WFC
0     2006-01-03  47.080002  492.899994  ...   40.189999  58.310001  31.900000
1     2006-01-04  46.580002  483.799988  ...   39.619999  58.349998  31.530001
2     2006-01-05  46.639999  486.200012  ...   39.740002  58.509998  31.495001
3     2006-01-06  46.570000  486.200012  ...   40.020000  58.570000  31.680000
4     2006-01-09  46.599998  483.899994  ...   40.669998  59.189999  31.674999
...          ...        ...         ...  ...         ...        ...        ...
3744  2020-11-16  27.580000   50.400002  ...  117.300003  58.110001  24.900000
3745  2020-11-17  27.549999   50.730000  ...  116.110001  58.810001  25.040001
3746  2020-11-18  26.980000   51.049999  ...  115.250000  58.810001  25.059999
3747  2020-11-19  26.980000   52.139999  ...  115.559998  58.630001  26.160000
3748  2020-11-20  26.809999   51.650002  ...  114.570000  57.810001  25.480000

[3749 rows x 7 columns]

# *Perform Random Asset Allocation And Calculate Portfolio Daily Return*

In [20]:
# Assuming I have 1 Million of dollars to invest
# Assuming the following fixed weights allocation
# ASSETS = [BAC  C  GS  JPM  MS  WFC]
# WEIGHTS = [10.92%  12.06%  0.602%  3.627%  14.492%  17.636%]
# Let's create random portfolio weights
# Portfolio weights must sum to 1 
# Set random seed
# np.random.seed(101)
np.random.seed()

In [21]:
# Create random weights for the stocks and normalize them
weights = np.array(np.random.random(6))
weights

array([0.43495941, 0.0211645 , 0.53208081, 0.51815079, 0.13789465,
       0.47425097])

In [22]:
# Ensure that the sum of all weights are = 1
weights = weights / np.sum(weights) 
print(weights)

[0.20531469 0.00999032 0.25115909 0.24458367 0.06509067 0.22386156]


In [23]:
# Function to normalize stock prices based on their initial price 
def normalize(df):
  x = df.copy()
  for i in x.columns[1:]:
    x[i] = x[i]/x[i][0]
  return x

In [24]:
# Normalize the stock values 
df_portfolio = normalize(bank_stocks_close)
df_portfolio

DATE       BAC         C        GS       JPM        MS       WFC
0     2006-01-03  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000
1     2006-01-04  0.989380  0.981538  0.986188  0.985817  1.000686  0.988401
2     2006-01-05  0.990654  0.986407  0.985800  0.988803  1.003430  0.987304
3     2006-01-06  0.989167  0.986407  0.999767  0.995770  1.004459  0.993103
4     2006-01-09  0.989805  0.981741  1.011795  1.011943  1.015092  0.992947
...          ...       ...       ...       ...       ...       ...       ...
3744  2020-11-16  0.585811  0.102252  1.725615  2.918637  0.996570  0.780564
3745  2020-11-17  0.585174  0.102921  1.743230  2.889027  1.008575  0.784953
3746  2020-11-18  0.573067  0.103571  1.740514  2.867629  1.008575  0.785580
3747  2020-11-19  0.573067  0.105782  1.742609  2.875342  1.005488  0.820063
3748  2020-11-20  0.569456  0.104788  1.733142  2.850709  0.991425  0.798746

[3749 rows x 7 columns]

In [25]:
df_portfolio.columns[1:]

Index(['BAC', 'C', 'GS', 'JPM', 'MS', 'WFC'], dtype='object')

In [26]:
# Note that enumerate returns the value and a counter as well
for counter, stock in enumerate(df_portfolio.columns[1:]):
  df_portfolio[stock] = df_portfolio[stock] * weights[counter]
  df_portfolio[stock] = df_portfolio[stock] * 1000000
df_portfolio

DATE            BAC  ...            MS            WFC
0     2006-01-03  205314.694349  ...  65090.665425  223861.562871
1     2006-01-04  203134.207026  ...  65135.313648  221265.056910
2     2006-01-05  203395.854858  ...  65313.919312  221019.441843
3     2006-01-06  203090.587963  ...  65380.898033  222317.694823
4     2006-01-09  203221.411879  ...  66072.994450  222282.599308
...          ...            ...  ...           ...            ...
3744  2020-11-16  120275.680422  ...  64867.407280  174738.335321
3745  2020-11-17  120144.848188  ...  65648.808659  175720.808974
3746  2020-11-18  117659.093970  ...  65648.808659  175861.150880
3747  2020-11-19  117659.093970  ...  65447.876754  183580.517886
3748  2020-11-20  116917.727948  ...  64532.522192  178808.545068

[3749 rows x 7 columns]

In [27]:
# Let's create an additional column that contains the sum of all $ values in the portfolio
df_portfolio['portfolio daily worth in $'] = df_portfolio[df_portfolio != 'DATE'].sum(axis = 1)
df_portfolio

DATE            BAC  ...            WFC  portfolio daily worth in $
0     2006-01-03  205314.694349  ...  223861.562871                1.000000e+06
1     2006-01-04  203134.207026  ...  221265.056910                9.881453e+05
2     2006-01-05  203395.854858  ...  221019.441843                9.890214e+05
3     2006-01-06  203090.587963  ...  222317.694823                9.952934e+05
4     2006-01-09  203221.411879  ...  222282.599308                1.003011e+06
...          ...            ...  ...            ...                         ...
3744  2020-11-16  120275.680422  ...  174738.335321                1.508158e+06
3745  2020-11-17  120144.848188  ...  175720.808974                1.506980e+06
3746  2020-11-18  117659.093970  ...  175861.150880                1.498725e+06
3747  2020-11-19  117659.093970  ...  183580.517886                1.508678e+06
3748  2020-11-20  116917.727948  ...  178808.545068                1.493837e+06

[3749 rows x 8 columns]

In [28]:
# Let's calculate the portfolio daily return 
# Define a new column in the dataframe and set it to zeros
df_portfolio['portfolio daily % return'] = 0.0000

for i in range(1, len(bank_stocks_close)):
  # Calculate the percentage of change from the previous day
  df_portfolio['portfolio daily % return'][i] = ( (df_portfolio['portfolio daily worth in $'][i] - df_portfolio['portfolio daily worth in $'][i-1]) / df_portfolio['portfolio daily worth in $'][i-1]) * 100 

df_portfolio

DATE  ...  portfolio daily % return
0     2006-01-03  ...                  0.000000
1     2006-01-04  ...                 -1.185473
2     2006-01-05  ...                  0.088666
3     2006-01-06  ...                  0.634164
4     2006-01-09  ...                  0.775424
...          ...  ...                       ...
3744  2020-11-16  ...                  2.451728
3745  2020-11-17  ...                 -0.078121
3746  2020-11-18  ...                 -0.547766
3747  2020-11-19  ...                  0.664118
3748  2020-11-20  ...                 -0.983718

[3749 rows x 9 columns]

- run the code 3 times, and record the final portolio value on the last day (November 20th, 2020)

- np.random.seed()


- Portfolio Value Run #1 = $1.228323e+06, Weights = [0.18523436 0.22302747 0.18397374 0.21051115 0.11913713 0.07811615]

- Portfolio Value Run #2 = $1.399220e+06, Weights = [0.21824053 0.07628536 0.05575036 0.30765668 0.10411314 0.23795392]

- Portfolio Value Run #3 = $1.493837e+06, Weights = [0.20531469 0.00999032 0.25115909 0.24458367 0.06509067 0.22386156]

# *Portfolio Allocation - Daily Return / Worth Calculation (Function)*

In [29]:
# Assuming I have $1,000,000 to be invested and I will allocate this fund based on the weights of the stocks
# I will create a function that takes in the stock prices along with the weights and return:
# (1) Daily value of each individual securuty in $ over the specified time period
# (2) Overall daily worth of the entire portfolio 
# (3) Daily return 

def portfolio_allocation(df, weights):

  df_portfolio = df.copy()
  
  # Normalize the stock avalues 
  df_portfolio = normalize(df_portfolio)
  
  for counter, stock in enumerate(df_portfolio.columns[1:]):
    df_portfolio[stock] = df_portfolio[stock] * weights[counter]
    df_portfolio[stock] = df_portfolio[stock] * 1000000

  df_portfolio['portfolio daily worth in $'] = df_portfolio[df_portfolio != 'DATE'].sum(axis = 1)
  
  df_portfolio['portfolio daily % return'] = 0.0000

  for i in range(1, len(bank_stocks_close)):
    
    # Calculate the percentage of change from the previous day
    df_portfolio['portfolio daily % return'][i] = ( (df_portfolio['portfolio daily worth in $'][i] - df_portfolio['portfolio daily worth in $'][i-1]) / df_portfolio['portfolio daily worth in $'][i-1]) * 100 
  
  # set the value of first row to zero, as previous value is not available
  df_portfolio['portfolio daily % return'][0] = 0
  return df_portfolio

In [30]:
# Call the function
df_portfolio = portfolio_allocation(bank_stocks_close, weights)
df_portfolio

DATE  ...  portfolio daily % return
0     2006-01-03  ...                  0.000000
1     2006-01-04  ...                 -1.185473
2     2006-01-05  ...                  0.088666
3     2006-01-06  ...                  0.634164
4     2006-01-09  ...                  0.775424
...          ...  ...                       ...
3744  2020-11-16  ...                  2.451728
3745  2020-11-17  ...                 -0.078121
3746  2020-11-18  ...                 -0.547766
3747  2020-11-19  ...                  0.664118
3748  2020-11-20  ...                 -0.983718

[3749 rows x 9 columns]

# *Perform Portfolio Data Visualization*

In [31]:
# Plot the portfolio daily return
# Zoom in to view
fig = px.line(x = df_portfolio.DATE, y = df_portfolio['portfolio daily % return'], title = 'Portfolio Daily % Return')
fig.show()

In [32]:
# Function to plot interactive plots using Plotly Express
def interactive_plot(df, title):
  fig = px.line(title = title)
  for i in df.columns[1:]:
    fig.add_scatter(x = df['DATE'], y = df[i], name = i)
  fig.show()

In [33]:
# Plot all stocks (normalized)
interactive_plot(df_portfolio.drop(['portfolio daily worth in $', 'portfolio daily % return'], axis = 1), 'Portfolio individual stocks worth in $ over time')

Answer to question 5: The following Banks have dropped on January 20th, 2009 (Zoom in to view)

- WFC with 88.17434k
- JPM with 66.27827k
- MS with 44.62895k
- GS with 36.39928k
- C with 2,706.285

Answer to question 4: Let's select the dates each Bank had the best and worst single day returns

- BAC had the best day return on November 2006; and the worst day return on March 2009
- C had the best day return on December 2006; and the worst day return on November 2020
- GS had the best day return on March 2018; and the worst day return on December 2008
- JPM had the best day return on January 2020; and the worst day return on January 2009
- MS had the best day return on June 2007; and the worst day return on October 2008
- WFC had the best day return on January 2018; and the worst day return on March 2009

In [34]:
# Print out a histogram of daily returns
fig = px.histogram(df_portfolio, x = 'portfolio daily % return')
fig.show()

In [35]:
# Plot the portfolio overall daily worth vs. time
fig = px.line(x = df_portfolio.DATE, y = df_portfolio['portfolio daily worth in $'], title = 'Portfolio Overall Value in $')
fig.show()

# *Calculate Portfolio Statistical Metrics (Cummulative Return, Average Daily Return, And Sharpe Ratio)*

- Stock daily return is a calculation of how much investors have gained or lost per day
- Cumulative return is a measure of the aggregate amount that the stock gained or lost over a period of time
- The standard deviation is a measurement of the dispersion away from the mean
- Sharpe ratio is used by investors to calculate the return of an investment compared to its risk
- Sharpe ratio is simply a calculation of the average return earned in excess of the risk free rate per unit of risk.

In [36]:
df_portfolio

DATE  ...  portfolio daily % return
0     2006-01-03  ...                  0.000000
1     2006-01-04  ...                 -1.185473
2     2006-01-05  ...                  0.088666
3     2006-01-06  ...                  0.634164
4     2006-01-09  ...                  0.775424
...          ...  ...                       ...
3744  2020-11-16  ...                  2.451728
3745  2020-11-17  ...                 -0.078121
3746  2020-11-18  ...                 -0.547766
3747  2020-11-19  ...                  0.664118
3748  2020-11-20  ...                 -0.983718

[3749 rows x 9 columns]

In [37]:
# Cumulative return of the portfolio (Note that I now look for the last net worth of the portfolio compared to it's start value)
cumulative_return = ((df_portfolio['portfolio daily worth in $'][-1:] - df_portfolio['portfolio daily worth in $'][0])/ df_portfolio['portfolio daily worth in $'][0]) * 100
print('Cumulative return of the portfolio is {} %'.format(cumulative_return.values[0]))

Cumulative return of the portfolio is 49.38369675592049 %


In [38]:
# Calculate the portfolio standard deviation
print('Standard deviation of the portfolio is {}'.format(df_portfolio['portfolio daily % return'].std()))

Standard deviation of the portfolio is 2.3387016221525916


In [39]:
# Calculate the average daily return 
print('Average daily return of the portfolio is {} %'.format(df_portfolio['portfolio daily % return'].mean() ))

Average daily return of the portfolio is 0.03784212516312053 %


In [40]:
# Portfolio sharpe ratio
sharpe_ratio = df_portfolio['portfolio daily % return'].mean() / df_portfolio['portfolio daily % return'].std() * np.sqrt(252)
print('Sharpe ratio of the portfolio is {}'.format(sharpe_ratio))

Sharpe ratio of the portfolio is 0.25686265742179565
